torch/csrc/Stream.cpp

```diff
@@ -37,9 +72,27 @@ static PyObject* THPStream_pynew(
   }
 
   THPStream* self = (THPStream*)ptr.get();
-  self->stream_id = stream_id;
-  self->device_index = device_index;
-  self->device_type = device_type;
+
+  // If torch.Stream is not created from existing Stream, retrieve one from the
+  // stream pool. It requires other device backends override
+  // getStreamFromGlobalPool method.
```

Comment:
getStream is used to track current stream used by this thread. getStreamFromGlobalPool returns pre-created stream in the pool. I currently feel it is actually better to use getStreamFromGlobalPool. Since streams are typically moving around and copied to another stream object, letting the pool manage the lifetime of those streams make the implementation much simpler.

Reply:
I agree we need to separate the concepts of
- Get the currently active stream for the currently active device/device_idx
- Get a fresh new stream

I guess the proposal here to use getStream/getStreamFromGlobalPool is a close analogous to the cuda naming we have today.
tbh I think a clearer naming would be good here since we have the luxury to be able to rename these APIs. In particular getCurrentStream/getNewStream sound simpler to understand to me (the fact that the new stream comes from a pool or not is an implementation detail).

Reply:
vote for adding a getNewStream API.